Fix/handle async failed refunds

achyuthajoy opened the pull request:
Fixes #2285 

#### Changes proposed in this Pull Request

At the moment when a WCPay refund fails and the webhook request is received, we update the order meta `_wcpay_refund_status` to `failed`. However, the refund is not deleted. This PR enables fetching the last WC Refund and saving the WCPay refund id as meta, which is then used to delete WC Refunds when the failed webhook request is received.

This PR also updates the refund order notes to include the WCPay refund id and changing the order status to `failed` if the order is fully refunded. For partially refunded orders, the status is not changed.

#### Testing instructions

* Follow the steps mentioned in #2285 to replicate the problem.
* Switch to the branch `fix/handle-async-failed-refunds`.
* The changes can be tested locally, but the Jurassic Tube tunnel needs to be enabled to receive webhook requests.
* Create a new order from front-end with the card `4000 0000 0000 5126`
* Fully refund the order from WP-ADMIN and confirm that the order note added includes the WCPay refund id - `re_xxxxxxxxxxxxxxxxxxxxxxxx`
* Refresh the page to see the new order note indicating refund failure.
* Confirm that the refund create is no longer present on the order and the order status is changed to `Failed`
* Create another order using the same card details above and do a partial refund.
* Confirm that the refund create is no longer present on the order and the order status is the same as before.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : _Add link here / 'QA Testing Not Applicable'_

htdat commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1463,19 +1465,37 @@ public function process_refund( $order_id, $amount = null, $reason = '' ) {
+					]
+				),
+				WC_Payments_Explicit_Price_Formatter::get_explicit_price( wc_price( $amount, [ 'currency' => $currency ] ), $order ),
+				$refund['id']
```
Yup, great idea to add the refund ID for relevant notes.

htdat commented on includes/class-wc-payments-webhook-processing-service.php:
```diff
@@ -202,6 +202,17 @@ private function process_webhook_refund_updated( $event_body ) {
 		);
+
+		// Delete refund based on WCPay refund id.
+		$wc_refunds = $order->get_refunds();
```
Nitpick: to aid IDE

```suggestion
		/** @var $wc_refunds WC_Order_Refund[] */
		$wc_refunds = $order->get_refunds();
```

htdat reviewed (changes requested):
Great job in figuring out proper places to add the relevant actions. 

As discussed via DMs, the main purpose of introducing `_wcpay_active_refund_id`is that during process_refund , we can not know the current refund ID. 

However, the related WC_Order_Refund object should be created before `process_refund`, we can explore using the last refund. That would eliminate introducing `_wcpay_active_refund_id` and make the code change more clear.

htdat commented on includes/class-wc-payments-webhook-processing-service.php:
```diff
@@ -202,6 +202,17 @@ private function process_webhook_refund_updated( $event_body ) {
+			$wcpay_refund_id = $wc_refund->get_meta( '_wcpay_refund_id', true );
+			if ( $refund_id === $wcpay_refund_id ) {
+				$wc_refund->delete();
+				break;
```
@achyuthajoy - I am working on this issue https://github.com/automattic/woocommerce-payments/issues/3814#issuecomment-1066691553 and since you're working here, I think it would integrate it for handling `charge.refund.updated` event. 

More specifically, if there is no refund ID having the `refund_id` from the webhook data, let's stop right away without processing further.

achyuthajoy replied:
Sure @dat.

Just wanted to confirm that we need to skip processing the refund event, if `_wcpay_refund_status` is already marked as failed or if we can't find a refund associated with the wcpay refund id. Correct?

htdat replied:
Thank you. 

> if we can't find a refund associated with the wcpay refund id.

It would be only this condition :)

htdat replied:
After discussing further, we decided to withdraw this idea and use the same approach `order_note_exists` in this PR https://github.com/Automattic/woocommerce-payments/pull/3975

Going to mark this resolved.

achyuthajoy commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2710,4 +2730,23 @@ function( $cost, $qty ) {
+		$order = wc_get_order( absint( $parent_order_id ) );
+		if ( ! is_object( $order ) ) {
+			return;
+		}
```
Code updated. Removed object check.

achyuthajoy commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -351,6 +351,8 @@ public function __construct(
 		// Update the current request logged_in cookie after a guest user is created to avoid nonce inconsistencies.
 		add_action( 'set_logged_in_cookie', [ $this, 'set_cookie_on_current_request' ] );
+
+		add_action( 'woocommerce_after_order_refund_object_save', [ $this, 'add_refund_id_to_order_meta' ], 10, 2 );
```
@htdat This action is triggered here - https://github.com/woocommerce/woocommerce/blob/9eca3a58250ac3b03f47dc26677c5539b1b653ae/plugins/woocommerce/includes/abstracts/abstract-wc-order.php#L200

I've removed the usage of this action based on the new approach.

htdat replied:
Gotcha! Thank you. It's really hard to catch up for this kind of variable action name.

achyuthajoy commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1463,19 +1465,37 @@ public function process_refund( $order_id, $amount = null, $reason = '' ) {
+		// Get active WC refund id from order and save WCPay refund id as meta.
+		$active_refund_id = $order->get_meta( '_wcpay_active_refund_id', true );
+		if ( ! empty( $active_refund_id ) ) {
+			update_post_meta( $active_refund_id, '_wcpay_refund_id', $refund['id'] );
```
Code updated to use built-in methods

achyuthajoy commented:
@htdat I've added the unit tests for the changes. Based on the discussion (p1648029332849129-slack-CGGCLBN58) regarding setting order status, I've added commits to update the order status after deleting the refund in question.

Ready for your review again :)

achyuthajoy commented:
Testing instructions updated.

htdat commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1485,19 +1486,39 @@ public function process_refund( $order_id, $amount = null, $reason = '' ) {
 		}
 
+		// Get the last created WC refund from order and save WCPay refund id as meta.
+		$wcpay_db       = new WC_Payments_DB();
+		$wc_last_refund = $wcpay_db->last_refund_from_order_id( $order_id );
```
I think it would be better to move the util method `last_refund_from_order_id` to a public static method in `WC_Payments_Utils` due to 2 reasons: 

- Initializing this service `WC_Payments_DB` is a bit contrast with our effort of currently using it and plug into it the relevant services in the entry class `WC_Payments`. 
- The method itself is not tightly coupled with WCPay, and it can be used by anywhere else in our codebase. Hence, a public static function `WC_Payments_Utils` is a good fit.

htdat commented on includes/class-wc-payments-db.php:
```diff
@@ -110,4 +110,24 @@ private function order_id_from_meta_key_value( $meta_key, $meta_value ) {
+	 *
+	 * @param string $order_id WC Order ID.
+	 *
+	 * @return bool|WC_Order|WC_Order_Refund
```
To ease the usage of this util function, returning either `null` or `WC_Order_Refund` seems a better option. 

Also, I think we may try our best to avoid using the direct SQL command. IMHO, `wc_get_orders` is still a good choice to use here. It provides [enough arguments (limit, orderby, and type)](https://github.com/woocommerce/woocommerce/wiki/wc_get_orders-and-WC_Order_Query) to get the same result and ensure the compatibility for [the future of a custom order table](https://developer.woocommerce.com/2022/01/17/the-plan-for-the-woocommerce-custom-order-table/).

htdat replied:
We may add tests to cover this util function too.

achyuthajoy replied:
Updated the PHPdoc comment and replaced DB query with `wc_get_orders`. I've originally added a [commit](https://github.com/Automattic/woocommerce-payments/pull/3965/commits/a3a8935376de6c56338f804b0164b2f917e049d1) that uses `wc_get_orders` and dropped it to use DB query.

htdat reviewed (commented):
✅ Thanks for adding solid tests.
✅ Manual tests look great. 

💬 I've left some comments to improve the util function and the logic of updating order status a bit.

achyuthajoy commented on includes/class-wc-payments-webhook-processing-service.php:
```diff
@@ -202,6 +202,30 @@ private function process_webhook_refund_updated( $event_body ) {
+					$current_status = $order->get_status();
+					if ( 'refunded' === $current_status ) {
+						$order->update_status( 'failed' );
+					}
```
Moved status change logic outside loop.

achyuthajoy commented:
That for the detailed reviews @htdat 👍
